### ISSUE TITLE
Regression detected with #11059

### ISSUE THREAD

**Issue by beta-ziliani** [crystal-lang/crystal#11263]
The annotation inserted in the block of `Mutable#map!(& : T -> T)` is breaking [mint-lang](https://github.com/mint-lang/mint/runs/3716944528). If I remove it `Mutable#map!(&)` specs pass. I can't find yet an explanation, nor a reduced form of the code triggering the issue (working on it).

**Comment by HertzDevil** [comment]
```crystal
class Foo
  getter foo = [] of Foo | Bar
end

class Bar
end

def bar(type)
  type.foo.map! do |x| # Error: type must be Bar, not (Bar | Foo)
    if x.is_a?(Foo)
      bar(x)
    else
      x
    end
  end

  type
end

bar(Foo.new)
```

It seems to happen because of the recursive `bar` call.

**Comment by asterite** [comment]
It's probably a bug that was always there. It's probably better to revert that change and fix it for 1.3 if it's tricky to fix.

My guess is that a type is deduce for U and frozen, but later it changes and it breaks. It should probably not be frozen when it's deduced. Not sure!

**Comment by asterite** [comment]
Further reduced with this that works with `--prelude=empty`:

```crystal
class Gen(T)
  def initialize(@value : T)
  end

  def map!(&block : T -> T) : self
    Gen(T).new(yield @value)
  end
end

def bar
  Gen.new(1 || 'a').map! do |x|
    if x.is_a?(Char)
      bar
    else
      x
    end
  end

  'a'
end

bar
```

**Comment by asterite** [comment]
The issue seems to be that when we analyze `T -> T`, we restrict the inferred block's type to `T`. So it will first start as `Int32` in the example above because the type of the recursive `bar` call is not yet known. `Int32` restricted to `Int32 | Char` is `Int32`. Later the block's type becomes `Int32 | Char` but that doesn't match `Int32`.

Ideally, when there's a restriction that returns `T`, the block's type must be compared against the actual `T`, in this case `Int32 | Char`, not the block's type restricted to `T`.

But this is a bit tricky when free variables are involved.

I don't have time to fix this now, but I hope someone will be able to do something with this information 😄

**Comment by beta-ziliani** [comment]
Thanks @HertzDevil and @asterite! I will update the ticket to reflect its true nature

**Comment by beta-ziliani** [comment]
FTR, if `bar` is annotated with the return type (`Foo` in the first example, `Char` in the second), it works. This supports your claim, @asterite .

**Comment by straight-shoota** [comment]
It appears that a simple user-space workaround would be adding a type restriction to the recursive method (or the block body).

Working around this issue in the method def as in #11267 has some issues and I doubt they can be mitigated to an acceptable state.

**Comment by HertzDevil** [comment]
It doesn't even need generics:

```crystal
def foo(& : Int32 | Char -> Int32 | Char)
end

def bar
  foo do |x| # Error: type must be Int32, not (Char | Int32)
    if x.is_a?(Char)
      bar
    else
      x
    end
  end

  'a'
end

bar
```

So it looks like we cannot introduce _any_ new block return type restrictions at all until this is fixed.

**Comment by asterite** [comment]
Right. Here too, the block type must be freezed to the specified block return type instead of the inferred type.
